### PR TITLE
update go to 1.17

### PR DESCRIPTION
**Pull request by thaJeztah**
carries / closes https://github.com/docker/cli/pull/3258

Removes the platform based switch between different versions.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,25 +1,14 @@
...
-FROM golatest AS go-windows-arm64
-FROM go-windows-${TARGETARCH} AS go-windows
-
 FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
 
-FROM go-${TARGETOS} AS build-base-alpine
+FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-${BASE_VARIANT} AS build-base-alpine
 COPY --from=xx / /
 RUN apk add --no-cache bash clang lld llvm file git
 WORKDIR /go/src/github.com/docker/cli
```
Comment (thaJeztah) [inline Dockerfile:14]: I was looking at these stages, and it appears we're not using the alpine stages anywhere? (perhaps I overlooked though 🤔)

**Comment by thaJeztah** [comment]
Looks like this is failing;

```
#18 https://github.com/theupdateframework/notary/releases/download/v0.6.1/notary-Linux-amd64
#18 ERROR: invalid response status 503
```

**Comment by thaJeztah** [comment]
okay; this looks ✅ now

@crazy-max @silvin-lubecki PTAL (please double check if I didn't miss something in the Dockerfile changes)

**Comment by thaJeztah** [comment]
```dockerfile
FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-${BASE_VARIANT} AS build-base-buster
```

Actually looking now if we need to use `BASE_VARIANT` in that one, as that would make the alpine and buster ones the same. (We probably _do_ need it to pick the right base though).

Also; guess we should be updating to `bullseye` now

**Comment by crazy-max** [comment]
@thaJeztah `libgcc-8-dev` > `libgcc-10-dev` or `libgcc-*-dev`.

**Comment by thaJeztah** [comment]
🤦 thought I could just a quick find/replace. Of course I could expect that 😅

**Comment by thaJeztah** [comment]
Hmmm... this looks like something we need to fix; are we missing the `osusergo` build-tag? (or do we need to compile without cgo here?) see https://github.com/golang/go/issues/23265

```
#53 [linux/amd64 build 2/2] RUN --mount=type=bind,target=.,ro     --mount=type=cache,target=/root/.cache     --mount=from=dockercore/golang-cross:xx-sdk-extras,target=/xx-sdk,src=/xx-sdk     --mount=type=tmpfs,target=cli/winresources     xx-go --wrap &&     TARGET=/out ./scripts/build/binary &&     xx-verify $([ "static" = "static" ] && echo "--static") /out/docker
#53 953.6 # github.com/docker/cli/cmd/docker
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000022.o: in function `New':
#53 953.6 /go/src/github.com/docker/cli/vendor/github.com/miekg/pkcs11/pkcs11.go:75: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000018.o: in function `mygetgrouplist':
#53 953.6 /usr/local/go/src/os/user/getgrouplist_unix.go:18: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000017.o: in function `mygetgrgid_r':
#53 953.6 /usr/local/go/src/os/user/cgo_lookup_unix.go:40: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000017.o: in function `mygetgrnam_r':
#53 953.6 /usr/local/go/src/os/user/cgo_lookup_unix.go:45: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000017.o: in function `mygetpwnam_r':
#53 953.6 /usr/local/go/src/os/user/cgo_lookup_unix.go:35: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000017.o: in function `mygetpwuid_r':
#53 953.6 /usr/local/go/src/os/user/cgo_lookup_unix.go:30: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#53 953.6 /usr/bin/s390x-linux-gnu-ld: /tmp/go-link-2402600021/000004.o: in function `_cgo_3c1cec0c9a4e_C2func_getaddrinfo':
#53 953.6 /tmp/go-build/cgo-gcc-prolog:58: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

**Comment by thaJeztah** [comment]
Actual _failure_ is in this bit though;

```
 > [linux/amd64 build 2/2] RUN --mount=type=bind,target=.,ro     --mount=type=cache,target=/root/.cache     --mount=from=dockercore/golang-cross:xx-sdk-extras,target=/xx-sdk,src=/xx-sdk     --mount=type=tmpfs,target=cli/winresources     xx-go --wrap &&     TARGET=/out ./scripts/build/binary &&     xx-verify $([ "static" = "static" ] && echo "--static") /out/docker:
#69 946.2 ld.lld: error: allocatestack.c:429:(.text+0x183C): unknown relocation (313) against symbol __default_pthread_attr_lock
#69 946.2 ld.lld: error: allocatestack.c:528:(.text+0x185C): unknown relocation (313) against symbol __static_tls_align_m1
#69 946.2 ld.lld: error: allocatestack.c:538:(.text+0x1898): unknown relocation (313) against symbol __static_tls_size
#69 946.2 ld.lld: error: allocatestack.c:525:(.text+0x18B4): unknown relocation (313) against symbol _dl_stack_flags
#69 946.2 ld.lld: error: pthread_create.c:875:(.text+0x1B7C): unknown relocation (313) against symbol __stack_chk_guard
#69 946.2 ld.lld: error: allocatestack.c:447:(.text+0x1BC8): unknown relocation (313) against symbol __static_tls_size
#69 946.2 ld.lld: error: allocatestack.c:457:(.text+0x1BDC): unknown relocation (313) against symbol __static_tls_align_m1
#69 946.2 ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
#69 946.2 clang: error: linker command failed with exit code 1 (use -v to see invocation)
#69 946.2 
```

**Comment by thaJeztah** [comment]
I removed the update to debian bullseye from this PR, and the other fixes (`osusergo`, updating `xx` have been moved separate and are merged; gonna move the bullseye update separate so that we can look how to fix that.

After this is merged, I'll also look at using `gofmt` for go1.17 (and probably update to go1.18); small steps..

**Review by crazy-max — APPROVED** [review]

**Comment by thaJeztah** [comment]
Let's get this one in; I'll open a PR for the debian update, and will have a look at updating `gofmt`, and go1.18

**Comment by thaJeztah** [comment]
opened https://github.com/docker/cli/pull/3493 for the debian step

**Pull request merged**